src/main/java/catan/services/impl/PlayServiceImpl.java

```diff
@@ -243,84 +192,81 @@ public void buildSettlement(UserBean user, String gameIdString, String nodeIdStr
 
     @Override
     public void endTurn(UserBean user, String gameIdString) throws PlayException, GameException {
-        Integer nextMoveNumber = null;
+        log.debug("User {} tries to end his turn of game id {}", user == null ? "<EMPTY>" : user.getUsername(), gameIdString);
+
         GameBean game = gameUtil.getGameById(gameIdString, ERROR_CODE_ERROR);
-        List<List<String>> initialBuildingsSet = gameUtil.getInitialBuildingsSetFromJson(game.getInitialBuildingsSet());
+        validateGameStatusIsPlaying(game);
+        validateUserNotEmpty(user);
+        validateCurrentTurnOfUser(game, user);
+        validateActionIsAllowed(user, game, GameUserActions.END_TURN);
 
-        log.debug("User {} tries to end his turn of game id {}",
-                user == null ? "<EMPTY>" : user.getUsername(), gameIdString);
-        //TODO: move to common validation method
-        if (user == null) {
-            log.debug("User should not be empty");
-            throw new PlayException(ERROR_CODE_ERROR);
-        }
+        playUtil.updateNextMoveOrder(game);
+        playUtil.updateCurrentCycleBuildingNumber(game);
+        playUtil.updateAvailableUserActions(game);
 
-        if (game.getStatus() != GameStatus.PLAYING) {
-            log.debug("Cannot end turn of not playing game");
-            throw new GameException(ERROR_CODE_ERROR);
-        }
+        gameDao.updateGame(game);
+    }
 
-        //TODO: move to util method and refactor all other places
-        GameUserBean gameUserBean = null;
-        for (GameUserBean gameUser : game.getGameUsers()) {
-            if (gameUser.getUser().equals(user)) {
-                gameUserBean = gameUser;
-                break;
+    private void validateActionIsAllowed(UserBean user, GameBean game, GameUserActions requiredAction) throws PlayException {
+        AllAvailableActionsDetails actions = playUtil.getAllAvailableActions(getJoinedGameUser(user, game).getAvailableActions());
+        boolean isActionAllowed = false;
+        for (ActionDetails allowedActions : actions.getList()) {
```

Comment:
I don't like that we use DTO (AllAvailableActionsDetails ) in service logic, we need to think how to avoid this